AI Framework Foundation

charvimehradu opened the pull request:
Pull Request for AI Framework Feature.

### Summary of Changes

- `Response.php` is the foundational file for the AI framework that standardizes how responses are handled across all AI providers.
- It represents the results of an AI operation and will hold all the data returned from AI providers. 
- Each AI provider creates a response object using constructor and users interact with the response objects using methods.
- __get() method allows property-style access ($response->content).

09 June 2025
- `ProviderInterface.php`: All AI providers must implement these methods.
- `AbstractProvider.php`: All providers share this code. It contains shared functionalities for all providers. 

11 June 2025
- 'OpenAIProvider.php': Implementing OpenAIProvider core functionality
- The `Tests` folder contains basic php tests.

### Testing Instructions


### Documentation Changes Required

bembelimen commented on src/AbstractProvider.php:
```diff
@@ -0,0 +1,147 @@
+/**
+ * Part of the Joomla Framework AI Package
+ *
+ */
```
Please check the headers Joomla is using

bembelimen commented on src/AbstractProvider.php:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        $this->options = $options;
+        $this->httpFactory = $httpFactory ?: new HttpFactory();
```
```HttpFactory::getHttp()``` is a static method, so need to check if calling it in a object context raises any notice/errors.

bembelimen commented on src/AbstractProvider.php:
```diff
@@ -0,0 +1,147 @@
+     */
+    protected function validateResponse($response): bool
+    {
+        if ($response->code !== 200) {
```
Is only a 200 a valid response?

charvimehradu replied:
Thank you for reviewing the PR. I initially set it to only accept 200 because I was focused on the basic case, but this is too restrictive. I kept it default to 200 because only AI providers can create response objects. An AI provider will only decide if they pass or fail and set it as the correct status. I will update this, with the upcoming commits.

charvimehradu commented:
With the latest commit, the following changes have been made:

- Changed file headers to follow Joomla standards - Updated copyright notices and licensing information to match Joomla Framework conventions.
- Implemented OpenAIProvider core functionality - OpenAI provider implementation with default endpoint `v1/chat/completions` and default model `gpt-4o-mini`. Provider support is determined by checking for API key availability in environment variables. The prompt() and ask() methods build proper request payloads that convert user prompts into OpenAI's expected message format with role-based structure. It also implements `determineAIStatusCode()` method that maps OpenAI's finish_reason responses to status codes.
- HTTP errors are handled by AbstractProvider's validateResponse() method, while OpenAI-specific API errors are processed in OpenAIProvider's parseOpenAIResponse() method.
- `composer.json` created to handle Http framework dependency and testing purposes.
- Created three basic tests:
1. HTTP Integration Test (http_test.php) - Verifies Joomla HttpFactory integration, tests both static and instance method call. Instance method call passes the test.
2. Response Object Test (response_test.php) - Tests custom Response class creation.
3. OpenAI API Integration Test (api_test.php) - Tests API connectivity verification.

bembelimen commented on composer.json:
```diff
@@ -0,0 +1,32 @@
+    "require": {
+        "php": "^8.1.0",
+        "joomla/http": "dev-4.x-dev",
+        "joomla/filesystem": "~3.0"
```
Please use version 4.x now for both (they're released)
